# Javascript 에러 처리

## 에러처리란?
* 예상치 못하게 프로그램에 발생한 에러를 처리하기 위한 코드이다.
* 프로그램 실행중 특정 부분에서 에러가 발생하면 프로그램 전체가 멈추기 때문에, 이를 방지하고자 하는 것이다.
* 즉, 에러처리의 목적은 **에러가 발생해도 프로그램이 멈추지 않는 것!**
* 예전부터 try-catch문을 자주 봤었지만, 구체적으로 어떤 상황에 이걸 쓰는지 몰랐었다. TypeError, SyntaxError 등은 개발자가 개발 과정에서 잡을 수 있는데, 굳이 왜 try-catch문으로 감싸는지 몰랐는데, 오늘 부트캠프에서 에러처리를 하는 이유에 대해 배워서 기억하려고 기록한다.

## 에러처리 코드를 작성하는 경우
* 위에 쓴 것처럼, 개발자가 잡아낼 수 있는 에러는 개발과정에서 모두 처리하는게 맞다.
* 따라서, 실무에서 에러처리 코드를 작성해야하는 대표적인 예는 서드 파티 라이브러리/API(?) 등에서 예기치 못하게 에러가 발생해서 내가 작성한 프로그램에도 에러가 생기는 경우이다.
* 외부에서 날씨 정보/교통 정보 등을 받아오는 API를 내 프로그램의 특정 부분에서 사용하고 있는데, 그 API를 제공하는 서버에 문제가 생겼을 경우.. 에러처리를 하지 않으면 내 프로그램 전체가 멈춘다. 따라서, 이렇게 내가 작성한 코드와는 무관하게(예상치 못하게) 프로그램에 문제가 생기는 걸 방지해야 할 때. 그 부분에 try-catch문을 사용한다.
* 실무에서 이런 경우 이외에도 에러처리를 하는 부분이 있을텐데, 아직 경험이 부족하므로 개발 중 접하게되면 차차 추가해나가자.


## 에러처리 방법
### try-catch문

```js
try {
    // 에러가 발생할 가능성이 있는 코드를 여기에 작성한다.
} catch (error) {
    // try문에서 에러 발생 시 실행되는 코드

    // error 유형에 따라 분기해서 처리 가능

} finally {
    // try, catch 이후 !반드시! 실행됨
    // finally는 생략가능
}
```

#### try
* 예상치 못한 에러가 날 수 있는 코드 부분을 try문 안에 작성한다.
* try문 내부 코드를 실행하다가 에러가 발생하면, try문 실행을 멈추고 catch문으로 넘어간다.
#### catch
* try문에서 에러가 발생할 경우 실행되는 부분이다.
* 인자로 들어오는 error 객체의 유형에 따라 분기해서 처리할 수 있으나, 이 방법은 많이 쓰지 않는다.
-> instanceof
#### finally
* 에러 발생 여부에 상관없이 무조건 마지막에 한번 동작한다.
* 에러 여부와 상관없이 반드시 동작해야 하는 코드를 넣어주자.
* ex) 연결 끊는것, 초기화.. 등
